Mathlib/Topology/MetricSpace/Ultra/Basic.lean

```diff
@@ -0,0 +1,235 @@
+`Metric.isClosed_ball` and `Metric.isClosed_sphere` is proven in a later file
+using more conceptual results.
+
+TODO: Generalize to ultrametric uniformities
+
+## Tags
+
+ultrametric, nonarchimedean
+-/
+
+variable {X : Type*}
+
+/-- The `dist : X → X → ℝ` respects the ultrametric inequality
+of `dist(x, z) ≤ max (dist(x,y)) (dist(y,z))`. -/
+class IsUltrametricDist (X : Type*) [Dist X] : Prop where
+  dist_triangle_max : ∀ x y z : X, dist x z ≤ max (dist x y) (dist y z)
+
+open Metric
+
+variable [PseudoMetricSpace X] [IsUltrametricDist X] (x y z : X) (r s : ℝ)
+
+lemma dist_triangle_max : dist x z ≤ max (dist x y) (dist y z) :=
+  IsUltrametricDist.dist_triangle_max x y z
+
+namespace IsUltrametricDist
+
+lemma ball_eq_of_mem {x y : X} {r : ℝ} (h : y ∈ ball x r) : ball x r = ball y r := by
+  ext a
+  simp_rw [mem_ball] at h ⊢
+  constructor <;> intro h' <;>
+  exact (dist_triangle_max _ _ _).trans_lt (max_lt h' (dist_comm x _ ▸ h))
+
+lemma mem_ball_iff_mem_ball {x y: X} {r : ℝ} : y ∈ ball x r ↔ x ∈ ball y r := by
+  cases lt_or_le 0 r with
+  | inl hr =>
+    constructor <;> intro h <;>
+    rw [← ball_eq_of_mem h] <;>
+    simp [hr]
+  | inr hr => simp [ball_eq_empty.mpr hr]
+
+lemma closedBall_eq_of_mem {x y: X} {r : ℝ} (h : y ∈ closedBall x r) :
+    closedBall x r = closedBall y r := by
+  ext
+  simp_rw [mem_closedBall] at h ⊢
+  constructor <;> intro h' <;>
+  exact (dist_triangle_max _ _ _).trans (max_le h' (dist_comm x _ ▸ h))
+
+lemma mem_closedBall_iff_mem_closedBall {x y: X} {r : ℝ} :
+    y ∈ closedBall x r ↔ x ∈ closedBall y r := by
+  cases le_or_lt 0 r with
+  | inl hr =>
+    constructor <;> intro h <;>
+    rw [← closedBall_eq_of_mem h] <;>
+    simp [hr]
+  | inr hr => simp [closedBall_eq_empty.mpr hr]
+
+lemma ball_subset_trichotomy :
+    ball x r ⊆ ball y s ∨ ball y s ⊆ ball x r ∨ Disjoint (ball x r) (ball y s) := by
+  cases le_or_lt r s with
+  | inl h =>
+    cases le_or_lt s (dist x y) with
+    | inl h' =>
+      refine Or.inr (Or.inr ?_)
+      intro S hsx hsy z hz
+      suffices y ∈ ball x s by
+        refine h'.not_lt ?_
+        simpa [dist_comm] using this
+      have hx : x ∈ ball z s := by
+        rw [mem_ball_iff_mem_ball]
+        exact ball_subset_ball h (hsx hz)
+      rw [← ball_eq_of_mem hx, mem_ball_iff_mem_ball]
+      exact hsy hz
+    | inr h' =>
+      refine Or.inl ?_
+      rw [ball_eq_of_mem (mem_ball.mpr h')]
+      exact ball_subset_ball h
+  | inr h =>
+    cases le_or_lt r (dist x y) with
+    | inl h' =>
+      refine Or.inr (Or.inr ?_)
+      intro S hsx hsy z hz
+      suffices y ∈ ball x r by
+        refine h'.not_lt ?_
+        simpa [dist_comm] using this
+      have hx : x ∈ ball z r := by
+        rw [mem_ball_iff_mem_ball]
+        exact hsx hz
+      rw [← ball_eq_of_mem hx, mem_ball_iff_mem_ball]
+      exact ball_subset_ball h.le (hsy hz)
+    | inr h' =>
+      refine Or.inr (Or.inl ?_)
+      rw [← ball_eq_of_mem (mem_ball.mpr h')]
+      exact ball_subset_ball h.le
```

Comment:
I don't really see the point of this lemma, especially in light of `ball_eq_or_disjoint`, which can be proven directly much more simply (see below).
```suggestion
```